doc/sonic-sub-port-intf-hld.md

```diff
@@ -0,0 +1,499 @@
+
+# Scope
+A sub port interface is a logical interface that can be created on a physical port or a port channel.
+A sub port interface serves as an interface to either a .1D bridge or a VRF, but not both.
+This design focuses on the use case of creating a sub port interface on a physical port or a port channel and using it as a router interface to a VRF as shown in Fig. 1.
+
+![](https://github.com/wendani/SONiC/blob/sub_port_master/images/sub_interface_hld/sub_intf_rif.png "Fig. 1: Sub port router interface")
+__Fig. 1: Sub port router interface__
+
+Multiple L3 sub port interfaces, each characterized by a VLAN id in the 802.1q tag, can be created on a physical port or a port channel.
+Sub port interfaces attaching to the same physical port or port channel can interface to different VRFs, though they share the same VLAN id space and must have different VLAN ids.
+Sub port interfaces attaching to different physical ports or port channels can use the same VLAN id, even when they interface to the same VRF.
+However, there is no L2 bridging between these sub port interfaces; each sub port interface is considered to stay in a separate bridge domain.
+
+This design does NOT cover the case of creating and using sub port as a bridge port to a .1D bridge shown in Fig. 2.
+
+![](https://github.com/wendani/SONiC/blob/sub_port_master/images/sub_interface_hld/sub_intf_bridge_port.png "Fig. 2: Sub port bridge port (out of scope of this design")
+__Fig. 2: Sub port bridge port__
+
+# Acronyms
+| Acronym                  | Description                                |
+|--------------------------|--------------------------------------------|
+| VRF                      | Virtual route forward                      |
+| RIF                      | Router interface                           |
+
+# 1 Requirements
+
+Manage the life cycle of a sub port interface created on a physical port or a port channel and used as a router interface to a VRF:
+* Creation with the specified dot1q vlan id encapsulation and optional mtu
+* Runtime mtu change
+* Removal
+
+A sub port interface shall support the following features:
+* L3 forwarding (both unicast and multicast)
+* BGP
+* ARP
+* VRF
+* RIF counters
+* QoS setting inherited from parent physical port or port channel
+* mtu inherited from parent physical port or port channel
+* Per sub port interface admin status config
+
+# 2 Schema design
+
+We introduce a new table "SUB_INTERFACE" in the CONFIG_DB to host the attributes of a sub port interface.
+For APPL_DB and STATE_DB, we do not introduce new tables for sub port interfaces, but reuse existing tables to host sub port interface keys.
+
+## 2.1 Configuration
+### 2.1.1 config_db.json
+```
+"SUB_INTERFACE": {
+    "{{ port_name }}.{{ vlan_id }}": {
+        "admin_status" : "{{ adminstatus }}"
+    },
+    "{{ port_name }}.{{ vlan_id }}|{{ ip_prefix }}": {}
+},
+
+"VLAN": {
+    "Vlan{{ vlan_id }}": {
+        "vlanid" : "{{ vlan_id }}"
+    }
+},
```

Comment:
I think we should not create clan since the clan id in the subject interface is just a tag.  it does not need vlan object to be created

Reply:
Yes @lguohan, but I got the reply from @wendani in the WG call that VLAN object creation is required for some reason, @wendani if not required, please remove the vlan-id configuration for sub-port feature.

Reply:
This is what the current SAI specification requires, and how vendor SAI is implemented if already supporting sub port interface. https://github.com/Mellanox/SAI-Implementation/blob/master/mlnx_sai/src/mlnx_sai_rif.c#L636

What Guohan suggests is to change the SAI spec before we can make any update accordingly here.

Reply:
the sai spec is not correct, can you change that?  @wendani